File: 프로그래머스/3/132266. 부대복귀/부대복귀.java
import java.util.*;

class Solution {
    public int[] solution(int n, int[][] roads, int[] sources, int destination) {
        // 1. 그래프 생성
        List<List<Integer>> graph = new ArrayList<>();
        for (int i = 0; i <= n; i++) {
            graph.add(new ArrayList<>());
        }
        for (int[] road : roads) {
            graph.get(road[0]).add(road[1]);
            graph.get(road[1]).add(road[0]);
        }

        // 2. 목적지로부터 모든 지점까지의 최단 거리 배열 초기화
        int[] dist = new int[n + 1];
        Arrays.fill(dist, -1);

        // 3. 목적지 기준 BFS 탐색
        Queue<Integer> q = new LinkedList<>();
        q.add(destination);
        dist[destination] = 0;

        while (!q.isEmpty()) {
            int current = q.poll();

            for (int next : graph.get(current)) {
                // 아직 방문하지 않은 지역인 경우
                if (dist[next] == -1) {
                    dist[next] = dist[current] + 1;
                    q.add(next);
                }
            }
        }

        // 4. sources 각 대원들의 최단 복귀 시간 결과 배열 생성
        int[] answer = new int[sources.length];
        for (int i = 0; i < sources.length; i++) {
            answer[i] = dist[sources[i]];
        }

        return answer;
    }
}

File: 프로그래머스/3/132266. 부대복귀/부대복귀.js
function solution(n, roads, sources, destination) {
    // 1. 그래프 생성
    const graph = Array.from({ length: n + 1 }, () => []);

    for (const [a, b] of roads) {
        graph[a].push(b);
        graph[b].push(a);
    }

    // 2. 목적지로부터 모든 지점까지의 최단 거리 배열 초기화
    const dist = Array(n + 1).fill(-1);

    // 3. 목적지 기준 BFS
    const queue = [destination];
    let front = 0;

    dist[destination] = 0;

    while (front < queue.length) {
        const current = queue[front++];

        for (const next of graph[current]) {
            // 아직 방문하지 않은 지역인 경우
            if (dist[next] === -1) {
                dist[next] = dist[current] + 1;
                queue.push(next);
            }
        }
    }

    // 4. sources의 최단 거리 반환
    return sources.map(source => dist[source]);
}